src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,384 @@
+    mPassword = NULL;
+    mOAuth2Token = NULL;
+    
+    mImapServices = new Array();
+    mSmtpServices = new Array();
+    mPopServices = new Array();
+    
+    mIdentifier = NULL;
+    mImapServer = NULL;
+    mPopServer = NULL;
+    mSmtpServer = NULL;
+    mImapError = ErrorNone;
+    mPopError = ErrorNone;
+    mSmtpError = ErrorNone;
+    
+    mCurrentServiceIndex = 0;
+    mCurrentServiceTested = 0;
+}
+
+AccountValidator::AccountValidator()
+{
+    init();
+}
+
+AccountValidator::~AccountValidator()
+{
+    MC_SAFE_RELEASE(mEmail);
+    MC_SAFE_RELEASE(mUsername);
+    MC_SAFE_RELEASE(mPassword);
+    MC_SAFE_RELEASE(mOAuth2Token);
+    MC_SAFE_RELEASE(mImapServices);
+    MC_SAFE_RELEASE(mSmtpServices);
+    MC_SAFE_RELEASE(mPopServices);
+    MC_SAFE_RELEASE(mIdentifier);
+}
+
+void AccountValidator::start()
+{
+    if (setup()) test();
+    else mCallback->operationFinished(this);
+}
+
+void AccountValidator::operationFinished(Operation * op)
+{
+    opCompleted();
+}
+
+bool AccountValidator::setup()
+{
+    if (mEmail == NULL) {
+        if (mUsername == NULL) return false;
+        else mEmail = mUsername;
+    }else if (mUsername == NULL){
+        mUsername = mEmail;
+    }
+    
+    MailProvider *provider = MailProvidersManager::sharedManager()->providerForEmail(mUsername);
+    
+    if (provider == NULL) {
+        Array * components;
+        String * domain;
+    
+        components = mUsername->componentsSeparatedByString(MCSTR("@"));
+        if (components->count() < 2)
+            return false;
+    
+        domain = (String *) components->lastObject();
+        provider = VTMxRecordForHostname(domain);
+    }
+    
+    if (provider != NULL) {
+        mIdentifier = provider->identifier();
+    
+        //If no custom Services have been set look for some in the provider
+        
+        if (mImapServices->count() == 0) {
+            if (provider->imapServices()->count() > 0) {
+                mImapServices = provider->imapServices();
+            } else {
+                mImapError = ErrorNoValidServerFound;
+            }
+        }
+    
+        if (mPopServices->count() == 0) {
+            if (provider->popServices()->count() > 0) {
+                mPopServices = provider->popServices();
+            } else {
+                mPopError = ErrorNoValidServerFound;
+            }
+        }
+    
+        if (mSmtpServices->count() == 0) {
+            if (provider->smtpServices()->count() > 0) {
+                mSmtpServices = provider->smtpServices();
+            } else {
+                mSmtpError = ErrorNoValidServerFound;
+            }
+    }
+    }
+    
+    return (mImapServices->count() > 0 || mPopServices->count() > 0 || mSmtpServices->count() > 0);
+}
+
+void AccountValidator::opCompleted()
+{
+    if (mCurrentServiceTested == 0) {
+        mImapError = ((IMAPOperation *)mOperation)->error();
+        (mImapError == ErrorNone) ? mCurrentServiceTested ++ : mCurrentServiceIndex ++;
```

Comment:
Use if / else with brace instead.
You could also refactor some code: error = mImapError;
if (error == ErrorNone) {
  // next type of service.
  mCurrentServiceTested ++;
}
else {
  // test next server.
  mCurrentServiceIndex ++;
}